File: app/renderer/src/lib/transitions.ts
// transitions.ts — the PURE renderer-side transition catalogue, guards and op builder.
//
// The sidecar owns the render (`media_studio/features/transitions.py`: xfade for
// video, acrossfade for audio, one boundary per extra clip). This module owns what
// the USER picks, and the two things the picker is obliged to show them BEFORE they
// commit to a render:
//
//   1. the real output duration — a transition OVERLAPS, so the timeline gets
//      SHORTER than the sum of the clips (`transitionOutputMs`), unlike a join;
//   2. that the render always re-encodes (`transitionReencodeNote`) — xfade has to
//      decode and recomposite both sides of every boundary, so it can never be a
//      stream copy.
//
// It also mirrors the engine's hard precondition (`transitionBlocker`): every clip
// must OUTLAST the transition it hosts. Enforcing it here means the user is stopped
// in the picker with a named clip instead of after a failed render.
//
// PURITY: no React, no rpc, no DOM — every export is a pure function, so the picker
// component stays a thin render shell and this logic is covered in isolation.

import type { DirectorOp } from './rpc';

/** A transition style id — mirrors the sidecar `transitions.TRANSITION_STYLES` keys. */
export type TransitionStyleId =
  | 'circleClose'
  | 'circleOpen'
  | 'dissolve'
  | 'fadeBlack'
  | 'fadeWhite'
  | 'slideLeft'
  | 'slideRight'
  | 'wipeDown'
  | 'wipeLeft'
  | 'wipeRight'
  | 'wipeUp';

/** One pickable transition: the wire id plus how it is described to the user. */
export interface TransitionStyle {
  id: TransitionStyleId;
  /** The human name shown on the control (never the raw wire id). */
  label: string;
  /** One plain-language line describing what the viewer will see. */
  blurb: string;
}

/**
 * The pickable transitions, in the SAME sorted order as the sidecar's
 * `STYLE_IDS = tuple(sorted(TRANSITION_STYLES))`. Order and membership are pinned
 * by this module's unit test and cross-checked against the Python side by the
 * sidecar's `test_transition_ts_parity`, because a drift between the two lists
 * would be invisible at runtime until a render failed with "unknown style".
 */
export const TRANSITION_STYLES: readonly TransitionStyle[] = [
  { id: 'circleClose', label: 'Circle close', blurb: 'The next clip closes in from a shrinking circle.' },
  { id: 'circleOpen', label: 'Circle open', blurb: 'The next clip opens out from a growing circle.' },
  { id: 'dissolve', label: 'Cross dissolve', blurb: 'The two clips blend through each other. The safe default.' },
  { id: 'fadeBlack', label: 'Fade through black', blurb: 'Dips to black between the clips — reads as a scene break.' },
  { id: 'fadeWhite', label: 'Fade through white', blurb: 'Dips to white between the clips — brighter, more upbeat.' },
  { id: 'slideLeft', label: 'Slide left', blurb: 'The next clip pushes the current one off to the left.' },
  { id: 'slideRight', label: 'Slide right', blurb: 'The next clip pushes the current one off to the right.' },
  { id: 'wipeDown', label: 'Wipe down', blurb: 'The next clip sweeps in from the top edge.' },
  { id: 'wipeLeft', label: 'Wipe left', blurb: 'The next clip sweeps in from the right edge.' },
  { id: 'wipeRight', label: 'Wipe right', blurb: 'The next clip sweeps in from the left edge.' },
  { id: 'wipeUp', label: 'Wipe up', blurb: 'The next clip sweeps in from the bottom edge.' },
];

/** The least opinionated boundary treatment — matches the sidecar `DEFAULT_STYLE`. */
export const DEFAULT_TRANSITION_STYLE: TransitionStyleId = 'dissolve';

/** Default transition length in ms — matches the sidecar `DEFAULT_DURATION_MS`. */
export const DEFAULT_TRANSITION_MS = 500;
/** Floor in ms — below this a dissolve is indistinguishable from a hard cut. */
export const MIN_TRANSITION_MS = 100;
/** Ceiling in ms — past this the transition dominates the clips it joins. */
export const MAX_TRANSITION_MS = 5000;

/** The label for a style id, falling back to the raw id for an unknown one. */
export function transitionStyleLabel(id: TransitionStyleId): string {
  return TRANSITION_STYLES.find((style) => style.id === id)?.label ?? id;
}

/**
 * Coerce a requested length to whole milliseconds inside the sidecar's accepted
 * range. A non-finite value (an empty/garbled number input) falls back to the
 * default rather than propagating NaN into the op params.
 */
export function clampTransitionMs(ms: number): number {
  if (!Number.isFinite(ms)) return DEFAULT_TRANSITION_MS;
  return Math.max(MIN_TRANSITION_MS, Math.min(MAX_TRANSITION_MS, Math.trunc(ms)));
}

/**
 * The real output length: the sum of the clips MINUS one overlap per boundary.
 *
 * This is the number the picker shows, and it is the whole point of the control —
 * a join makes the timeline longer by the length of what you added, a transition
 * makes it shorter than that by the overlap. Mirrors the sidecar's
 * `transitions.total_duration_sec`.
 */
export function transitionOutputMs(clipDurationsMs: readonly number[], transitionMs: number): number {
  const total = clipDurationsMs.reduce((sum, ms) => sum + ms, 0);
  return total - Math.max(0, clipDurationsMs.length - 1) * transitionMs;
}

/**
 * Why this selection cannot be transitioned, or `null` when it can.
 *
 * Mirrors the engine's precondition (`transitions.xfade_offsets`): at least two
 * clips, and every clip strictly LONGER than the transition it must host — a clip
 * that is not would be wholly consumed by the overlap, which ffmpeg renders as
 * garbage rather than refusing. Checked here so the user is stopped in the picker,
 * with the offending clip named, rather than after a wasted render.
 */
export function transitionBlocker(
  clipDurationsMs: readonly number[],
  transitionMs: number,
): string | null {
  if (clipDurationsMs.length < 2) {
    return 'A transition needs at least two clips — there is no boundary to treat.';
  }
  const seconds = (ms: number) => (ms / 1000).toFixed(1);
  for (const [index, ms] of clipDurationsMs.entries()) {
    if (ms <= transitionMs) {
      return `Clip ${index + 1} (${seconds(ms)}s) is shorter than the ${seconds(transitionMs)}s transition it would host.`;
    }
  }
  return null;
}

/**
 * The render-cost disclosure shown beside the picker, or `''` when there is no
 * boundary yet (nothing to disclose). Kept as data, not a static string, so the
 * boundary count is always the real one.
 */
export function transitionReencodeNote(clipCount: number): string {
  if (clipCount < 2) return '';
  const boundaries = clipCount - 1;
  const word = boundaries === 1 ? 'boundary' : 'boundaries';
  return `Transitions re-encode: the video is recomposited at ${boundaries} transition ${word}, so this export takes longer than a plain join and cannot be a lossless copy.`;
}

/** The inputs a user's picker selection turns into a wire op. */
export interface TransitionOpInput {
  /** The op id (caller-owned so the plan's ids stay unique and deterministic). */
  id: string;
  /** Paths of the clips joined AFTER the current source, in order. */
  clips: readonly string[];
  style?: TransitionStyleId;
  durationMs?: number;
}

/**
 * Build a wire-valid `transition` op from a picker selection.
 *
 * `span` is null on purpose: a transition acts on the JUNCTION between clips, not
 * on a source range, and the sidecar validator deliberately keeps `transition` out
 * of its span-required set for exactly that reason.
 *
 * The `rationale` is derived LOCALLY and deterministically from the selection —
 * it is not model text, so the storyboard row reads as what the user chose.
 */
export function buildTransitionOp(input: TransitionOpInput): DirectorOp {
  const style = input.style ?? DEFAULT_TRANSITION_STYLE;
  const durationMs = clampTransitionMs(input.durationMs ?? DEFAULT_TRANSITION_MS);
  return {
    id: input.id,
    kind: 'transition',
    span: null,
    params: { clips: [...input.clips], style, durationMs },
    reversible: true,
    rationale: `${transitionStyleLabel(style)} · ${(durationMs / 1000).toFixed(1)}s`,
    status: 'planned',
    statusReason: null,
  };
}

File: sidecar/media_studio/features/transitions.py
"""Boundary TRANSITIONS between joined clips — dissolve, fade-through, wipe, slide.

Before this module every join in Reframe was a HARD CUT: ``director_op_engines``
wired ``join`` to ffmpeg's concat *filter*, which butts clip B onto the end of
clip A with nothing in between, and there was no other way to express a boundary
treatment anywhere in the product.

The primitive here is ffmpeg's ``xfade`` (video) paired with ``acrossfade``
(audio). The one fact that makes a transition structurally different from a join:

    join      -> output duration == sum(parts)          (clips are ADJACENT)
    transition-> output duration == sum(parts) - N*D     (clips OVERLAP by D)

so a transition can never be a re-timing of a concat — it composites two clips at
the same instant, for ``D`` seconds, at each boundary.

RE-ENCODE HONESTY (the thing not to paper over): ``xfade`` must DECODE both sides
of every boundary and emit new frames, so a transition render can never be a
stream copy. That cost is exposed as a first-class, testable string —
:func:`reencode_note` — rather than a comment, so the UI can show it to the user
before they commit to the render.

Two neighbouring paths are deliberately NOT changed by this module, because both
were measured and neither is a transition:

  * ``director_op_engines.build_join_argv`` ALREADY re-encodes (concat filter,
    ``-c:v libx264`` at ``:359-362``). Adding transitions therefore regresses
    nothing there — ``join`` was never a stream copy, so "every join now
    re-encodes" was never on the table.
  * ``reframe_multispeaker.build_concat_argv`` IS a real ``-c copy`` stream-copy
    stitch (``:721-750``), but it stitches per-segment crops of ONE source clip
    inside the active-speaker reframe pass. It does not go through ``join`` and
    nothing here touches it, so that fast path stays intact.

PURE: stdlib + :mod:`media_studio.ffmpeg` binary resolution only. Every function
below is a string/number builder — no subprocess, no filesystem. The engine
adapter that runs the argv lives in ``director_op_engines.make_transition_engine``.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any

from media_studio import ffmpeg as _ffmpeg

#: Wire/UI transition ids -> the literal ffmpeg ``xfade`` transition name.
#:
#: The keys are camelCase because they travel on the RPC wire and into the
#: renderer's TS union; the VALUES are ffmpeg's own lowercase names and must not
#: drift. Deliberately a CURATED subset of xfade's ~50 transitions: each one below
#: is a boundary treatment an editor actually reaches for, and each has been in
#: ffmpeg since ``xfade`` landed in 4.3 — so the set does not depend on a bleeding
#: -edge binary. ``dissolve`` maps to xfade's ``fade`` (the classic cross-dissolve);
#: xfade also has a *separate* filter literally named ``dissolve`` (a noise-mask
#: wipe), which is NOT what an editor means by "dissolve" — hence the remap.
TRANSITION_STYLES: dict[str, str] = {
    "dissolve": "fade",
    "fadeBlack": "fadeblack",
    "fadeWhite": "fadewhite",
    "wipeLeft": "wipeleft",
    "wipeRight": "wiperight",
    "wipeUp": "wipeup",
    "wipeDown": "wipedown",
    "slideLeft": "slideleft",
    "slideRight": "slideright",
    "circleOpen": "circleopen",
    "circleClose": "circleclose",
}

#: The style ids in a stable, sorted order (the renderer renders them in this
#: order, and a set-vs-list drift between the two sides would be invisible).
STYLE_IDS: tuple[str, ...] = tuple(sorted(TRANSITION_STYLES))

#: The style applied when an op omits ``params['style']`` — a cross-dissolve, the
#: least opinionated boundary treatment.
DEFAULT_STYLE = "dissolve"

#: Transition length when an op omits ``params['durationMs']``. 500ms is the
#: conventional dissolve length; short enough not to eat a short clip.
DEFAULT_DURATION_MS = 500
#: Floor: below ~100ms a dissolve is indistinguishable from a hard cut.
MIN_DURATION_MS = 100
#: Ceiling: 5s. Past this the transition dominates the clips it joins, and every
#: clip would have to be longer than the transition (see :func:`xfade_offsets`).
MAX_DURATION_MS = 5000

#: Output frame rate used to conform inputs when the caller does not supply one.
DEFAULT_FPS = 30


class TransitionError(ValueError):
    """Raised on an impossible transition request (unknown style, clip too short).

    A :class:`ValueError` subclass so callers may catch either. The engine adapter
    re-raises it as a ``DirectorEngineError``, which ``apply_plan`` records as the
    op's ``status="failed"`` + reason with an auto-rollback — never a silent no-op
    and never a corrupt render.
    """


def normalize_style(raw: Any) -> str:
    """Resolve a wire style id to its ffmpeg ``xfade`` transition name.

    ``None`` (the param was omitted) yields :data:`DEFAULT_STYLE`. Anything else
    that is not a known id is REJECTED rather than silently defaulted: a user who
    asked for a star wipe must not quietly receive a cross-dissolve and be left
    thinking that is what a star wipe looks like.
    """
    if raw is None:
        return TRANSITION_STYLES[DEFAULT_STYLE]
    if isinstance(raw, str) and raw in TRANSITION_STYLES:
        return TRANSITION_STYLES[raw]
    raise TransitionError(f"unknown transition style: {raw!r} (known: {', '.join(STYLE_IDS)})")


def normalize_duration_ms(raw: Any) -> int:
    """Coerce a requested transition length to a whole number of ms, CLAMPED.

    ``None`` yields :data:`DEFAULT_DURATION_MS`; a number is truncated to int and
    clamped into ``[MIN_DURATION_MS, MAX_DURATION_MS]``. A non-number is rejected.

    ``bool`` is excluded explicitly: it is an ``int`` subclass in Python, so a
    stray ``True`` would otherwise become ``1`` and then clamp up to the floor —
    a nonsense value laundered into a plausible one.
    """
    if raw is None:
        return DEFAULT_DURATION_MS
    if isinstance(raw, bool) or not isinstance(raw, (int, float)):
        raise TransitionError(f"transition duration must be a number of milliseconds, got {raw!r}")
    return max(MIN_DURATION_MS, min(MAX_DURATION_MS, int(raw)))


def xfade_offsets(durations_sec: Sequence[float], duration_sec: float) -> list[float]:
    """Start time of each ``xfade`` boundary, in output-timeline seconds.

    ``xfade`` takes an ``offset`` measured on ITS OWN first input. Chained, that
    first input is the running output of the previous stage, which has already
    lost ``duration_sec`` per boundary consumed. So for boundary ``k``::

        offset_k = sum(durations[0..k]) - (k + 1) * duration_sec

    Every clip must be strictly LONGER than the transition: a clip that is not
    would be wholly consumed by the overlap and drive the offset negative, which
    ffmpeg renders as garbage rather than an error. Rejected loudly instead.
    """
    if len(durations_sec) < 2:
        raise TransitionError(f"a transition needs at least two clips, got {len(durations_sec)}")
    for index, clip_sec in enumerate(durations_sec):
        if clip_sec <= duration_sec:
            raise TransitionError(
                f"clip {index} ({clip_sec:.3f}s) is shorter than the {duration_sec:.3f}s "
                "transition it must host (a clip must outlast its transition)"
            )
    offsets: list[float] = []
    running = 0.0
    for index in range(len(durations_sec) - 1):
        running += durations_sec[index]
        offsets.append(running - (index + 1) * duration_sec)
    return offsets


def total_duration_sec(durations_sec: Sequence[float], duration_sec: float) -> float:
    """Output duration: the sum of the parts MINUS one overlap per boundary.

    This is the single strongest non-no-op proof that a transition ran rather than
    a concat: concat sums, a transition sums-then-subtracts.
    """
    return sum(durations_sec) - max(0, len(durations_sec) - 1) * duration_sec


def video_out_label(count: int) -> str:
    """Filtergraph label carrying the final composited VIDEO for ``count`` inputs."""
    return f"[vx{count - 1}]"


def audio_out_label(count: int) -> str:
    """Filtergraph label carrying the final crossfaded AUDIO for ``count`` inputs."""
    return f"[ax{count - 1}]"


def _conform_chain(index: int, width: int, height: int, fps: int) -> str:
    """Per-input conform stage: geometry, SAR, frame rate, pixel format.

    ``xfade`` REFUSES inputs whose size, SAR, frame rate or pixel format differ —
    it is a per-pixel blend of two frames, so they must be the same shape. This
    stage is therefore not cosmetic: it is what lets a transition join arbitrary
    clips at all, the same tolerance ``build_join_argv``'s concat filter has.
    Letterboxes rather than distorts (``force_original_aspect_ratio=decrease`` +
    centred ``pad``), so a 4:3 insert keeps its geometry instead of being stretched.
    """
    return (
        f"[{index}:v]scale={width}:{height}:force_original_aspect_ratio=decrease,"
        f"pad={width}:{height}:-1:-1:color=black,setsar=1,fps={fps},format=yuv420p[tv{index}]"
    )


def build_transition_filtergraph(
    *,
    count: int,
    xfade_name: str,
    duration_sec: float,
    offsets: Sequence[float],
    width: int,
    height: int,
    fps: int,
) -> str:
    """Build the full ``-filter_complex`` graph: conform -> xfade chain -> acrossfade chain.

    Video and audio are chained in LOCKSTEP — one ``xfade`` and one ``acrossfade``
    per boundary, both shortening their stream by ``duration_sec`` — so A/V stay in
    sync across every boundary. ``acrossfade`` needs no offset: it always crossfades
    the end of its first input with the start of its second, which is exactly the
    junction ``xfade`` is treating.

    Known scope limit, stated here rather than in a trailing caveat: for
    ``fadeBlack``/``fadeWhite`` the VIDEO dips through a colour while the AUDIO
    still cross-fades directly. A true fade-through-silence would need a separate
    ``afade`` out/in pair; that is not built. Settling experiment: render a
    ``fadeBlack`` and confirm the audio at the midpoint is a blend of both clips
    rather than silence.
    """
    if len(offsets) != count - 1:
        raise TransitionError(f"offset count {len(offsets)} does not match {count} inputs (need {count - 1})")
    stages = [_conform_chain(index, width, height, fps) for index in range(count)]
    for index, offset in enumerate(offsets):
        video_in = "[tv0]" if index == 0 else f"[vx{index}]"
        stages.append(
            f"{video_in}[tv{index + 1}]xfade=transition={xfade_name}:"
            f"duration={duration_sec:.3f}:offset={offset:.3f}[vx{index + 1}]"
        )
    for index in range(count - 1):
        audio_in = "[0:a]" if index == 0 else f"[ax{index}]"
        stages.append(f"{audio_in}[{index + 1}:a]acrossfade=d={duration_sec:.3f}:c1=tri:c2=tri[ax{index + 1}]")
    return ";".join(stages)


def build_transition_argv(
    inputs: Sequence[str],
    out_path: str,
    *,
    style: Any,
    duration_ms: Any,
    durations_sec: Sequence[float],
    width: int,
    height: int,
    fps: int = DEFAULT_FPS,
    settings: Mapping[str, Any] | None = None,
) -> list[str]:
    """argv that joins ``inputs`` with a transition at every boundary.

    ``durations_sec`` must carry the probed duration of EVERY input (the offset
    math is meaningless without them). Encodes H.264/AAC at CRF 18 with an explicit
    ``-preset``/``-pix_fmt`` — the same rate-control discipline
    ``reframe_multispeaker.build_segment_argv`` uses, deliberately NOT the bare
    ``-c:v libx264`` (default CRF 23, no floor) that the older Director builders
    pass, because a transition is by definition a generational re-encode and should
    not also silently drop quality.
    """
    if len(durations_sec) != len(inputs):
        raise TransitionError(f"need one duration per clip: {len(durations_sec)} durations for {len(inputs)} clips")
    xfade_name = normalize_style(style)
    duration_sec = normalize_duration_ms(duration_ms) / 1000.0
    offsets = xfade_offsets(durations_sec, duration_sec)
    graph = build_transition_filtergraph(
        count=len(inputs),
        xfade_name=xfade_name,
        duration_sec=duration_sec,
        offsets=offsets,
        width=width,
        height=height,
        fps=fps,
    )
    argv = [_ffmpeg.ffmpeg_path(settings), "-hide_banner", "-nostdin", "-y"]
    for path in inputs:
        argv += ["-i", path]
    argv += [
        "-filter_complex",
        graph,
        "-map",
        video_out_label(len(inputs)),
        "-map",
        audio_out_label(len(inputs)),
        "-c:v",
        "libx264",
        "-preset",
        "medium",
        "-crf",
        "18",
        "-pix_fmt",
        "yuv420p",
        "-c:a",
        "aac",
        "-b:a",
        "192k",
        "-movflags",
        "+faststart",
        "-progress",
        "pipe:1",
        "-nostats",
        out_path,
    ]
    return argv


def reencode_note(count: int) -> str:
    """The user-facing cost disclosure for a transition render.

    A first-class, asserted string rather than a source comment, because the cost
    is the user's to accept: ``xfade`` composites two decoded frames per boundary,
    so this render cannot be a stream copy under any settings.
    """
    if count < 2:
        raise TransitionError(f"a transition needs at least two clips, got {count}")
    boundaries = count - 1
    word = "boundary" if boundaries == 1 else "boundaries"
    return (
        f"Transitions re-encode. xfade decodes and recomposites both clips at each of the "
        f"{boundaries} transition {word}, so unlike a plain join this render can never be a "
        f"stream copy — and the result is shorter than the sum of the clips by one overlap "
        f"at every {word if boundaries == 1 else 'boundary'}."
    )


__all__ = [
    "DEFAULT_DURATION_MS",
    "DEFAULT_FPS",
    "DEFAULT_STYLE",
    "MAX_DURATION_MS",
    "MIN_DURATION_MS",
    "STYLE_IDS",
    "TRANSITION_STYLES",
    "TransitionError",
    "audio_out_label",
    "build_transition_argv",
    "build_transition_filtergraph",
    "normalize_duration_ms",
    "normalize_style",
    "reencode_note",
    "total_duration_sec",
    "video_out_label",
    "xfade_offsets",
]
